pkg/serverconfig/config.go

```diff
@@ -117,8 +118,13 @@ func addClusterInfo(fs *flag.FlagSet, clusterInfo *ClusterInfo) {
 
 func addAuth(fs *flag.FlagSet, auth *Auth) {
 	// Assume "openshift" if config file is used.
-	fs.Set("k8s-auth", "openshift")
-	fs.Set("user-auth", "openshift")
+
+	// FIXME!!!
+	fmt.Printf("k8s-auth: %s\n", fs.Lookup("k8s-auth").Value.String())
+	fmt.Printf("user-auth: %s\n", fs.Lookup("user-auth").Value.String())
+
+	// fs.Set("k8s-auth", "openshift")
+	// fs.Set("user-auth", "openshift")
```

Comment:
@spadgett @jhadvig @christianvogt At the moment I have the problem, that when we use a local `ConsoleConfig`, this code here previously overrides the auth=disabled option which is set by `contrib/oc-environment.sh`.

Without commenting out this two lines I can fetch the `ConsoleConfig` with this command:

```bash
oc get -o 'template={{index .data "console-config.yaml"}}' -n openshift-console configmaps console-config
```

But the fetched `ConsoleConfig` has more "problems", because it tries set some TLS files which are only available on a cluster. As example here is what I get from my local CRC:

```yaml
apiVersion: console.openshift.io/v1
auth:
  clientID: console
  clientSecretFile: /var/oauth-config/clientSecret
  oauthEndpointCAFile: /var/default-ingress-cert/ca-bundle.crt
clusterInfo:
  consoleBaseAddress: https://console-openshift-console.apps-crc.testing
  masterPublicURL: https://api.crc.testing:6443
customization:
  branding: ocp
  documentationBaseURL: https://docs.openshift.com/container-platform/4.6/
kind: ConsoleConfig
monitoringInfo:
  alertmanagerPublicURL: https://alertmanager-main-openshift-monitoring.apps-crc.testing
  grafanaPublicURL: https://grafana-openshift-monitoring.apps-crc.testing
  prometheusPublicURL: https://prometheus-k8s-openshift-monitoring.apps-crc.testing
  thanosPublicURL: https://thanos-querier-openshift-monitoring.apps-crc.testing
providers: {}
servingInfo:
  bindAddress: https://[::]:8443
  certFile: /var/serving-cert/tls.crt
  keyFile: /var/serving-cert/tls.key
```

And here one from a shared cluster:

```yaml
apiVersion: console.openshift.io/v1
auth:
  clientID: console
  clientSecretFile: /var/oauth-config/clientSecret
  oauthEndpointCAFile: /var/default-ingress-cert/ca-bundle.crt
clusterInfo:
  consoleBaseAddress: https://console-openshift-console.apps.REMOVED.devcluster.openshift.com
  masterPublicURL: https://api.REMOVED.devcluster.openshift.com:6443
customization:
  branding: ocp
  documentationBaseURL: https://docs.openshift.com/container-platform/4.7/
kind: ConsoleConfig
monitoringInfo:
  alertmanagerPublicURL: https://alertmanager-main-openshift-monitoring.apps.REMOVED.devcluster.openshift.com
  grafanaPublicURL: https://grafana-openshift-monitoring.apps.REMOVED.devcluster.openshift.com
  prometheusPublicURL: https://prometheus-k8s-openshift-monitoring.apps.REMOVED.devcluster.openshift.com
  thanosPublicURL: https://thanos-querier-openshift-monitoring.apps.REMOVED.devcluster.openshift.com
providers: {}
servingInfo:
  bindAddress: https://[::]:8443
  certFile: /var/serving-cert/tls.crt
  keyFile: /var/serving-cert/tls.key
```

When I reduce the `ConsoleConfig` to this example:

```yaml
apiVersion: console.openshift.io/v1
kind: ConsoleConfig
customization:
  developerCatalog:
    categories:
    - id: java
      label: Java
      tags: [java, jvm]
```

I can start the bridge with this file:

```bash
export BRIDGE_CONFIG=consoleconfig-with-dev-catalog-categories.yaml
bin/bridge
```

Reply:
@spadgett @jhadvig @christianvogt Two ideas to solve the two problem above:

**Issue a) Could not run disabled authentification locally with a ConsoleConfig file**

Currently the [config.go addAuth method](https://github.com/openshift/console/blob/e92878c521f90cc09de714ef4b61840a45ec963e/pkg/serverconfig/config.go#L118-L121) overrides the flags `k8s-auth` and `user-auth`. I expect just removing this is currently a risk. The current  code assume "if there is any config, we use openshift for auth". There is currently no way to override this.

The reason for this is, in [main.go line 124-138](https://github.com/openshift/console/blob/e92878c521f90cc09de714ef4b61840a45ec963e/cmd/bridge/main.go#L124-L138) the order of reading configuration flags is:

1. Program arguments
2. Environment variables
3. Config file

As user of the bridge I had expected a revere order. So that config parameters are overridden by environment variables, and both can be overridden by program arguments.

If we would change this order, we could provide the customization (and all other ConsoleConfig arguments) as config file. This could then still enable "openshift" authentification by default.

But the exported environment variables from `oc-environment.sh` can override this, so that the authentication is disabled for local development.

What do you think?

**Issue b) Cluster ConsoleConfig YAML could not be used locally**

I talked with Christian about the problem, that the `ConsoleConfig` contains many keys which doesn't work locally. Esp. the TLS configuration. His recommendation is that we fetch the yaml file and accept only "well known" keys.

This must not be part of the `oc-environment.sh` script. For example, a script `contrib/oc-customization.sh` could download the yaml and accept only customization data and `export BRIDGE_CONFIG`.

What your opinion here? Have you another recommendation?